### ISSUE TITLE
nre bugs / missing features

### ISSUE THREAD

**Issue by timotheecour** [nim-lang/Nim#8150]
```
import os
import nre

proc replaceMatch(a:RegexMatch):auto=
  let temp = a.captures

  when false: #ERROR: $() not defined
    echo a.captures

  when true: #ERROR: `len` not defined
    echo a.captures.len

  return a.captures[0]

proc test()=
  let a = replace("foo", re"(\w+)", replaceMatch)

test()
```

- [ ] the doc https://nim-lang.org/docs/nre.html#RegexMatch for `RegexMatch` is a bit unclear on the fact that `pattern` etc are procs that take a `RegexMatch` as input (eg https://nim-lang.org/docs/re.html doesn't have that issue)

- [ ] echo a.captures doesn't work (could be fixed by https://github.com/nim-lang/Nim/issues/8149 which is a more general issue)

- [ ] a.captures.len is not defined

- [ ] not clear how to call replace to only operate on first match (as opposed to all matches)

NOTE: I know `re` exists and was meant as a replacement for `nre`, however I depend on functionality that is in `nre` and not in `re`, eg replace with a `proc`

**Comment by flaviut** [comment]
> the doc https://nim-lang.org/docs/nre.html#RegexMatch for RegexMatch is a bit unclear on the fact that pattern etc are procs that take a RegexMatch as input (eg https://nim-lang.org/docs/re.html doesn't have that issue)

True, but the idea here is that you shouldn't have to know how exactly those things work. `pattern` is in fact an attribute on `RegexMatch`; `captures` is in fact a function that returns a `Captures` object, which is just a distinct alias of `RegexMatch`.

The idea is that you look at the examples on how to use it, and trust that the implementation is as efficient as possible. `captures`, for example, could return a `seq[string]`, but it doesn't in order to do things as lazy as possible to improve efficiency. Why should you need to allocate & copy a half-dozen strings and a seq when you're only interested in one item?

>  echo a.captures doesn't work

Agreed. This would be nice.

>  a.captures.len is not defined

How would you like this defined? There are a couple ways:

- total number of captures (this is also defined in `Regex.captureCount`)
- number of matched captures (this isn't particularly useful, since some captures are optional)

>  not clear how to call replace to only operate on first match (as opposed to all matches)

True. There's no method dedicated to doing this, but you could use a closure to https://nim-lang.org/docs/nre.html#replace%2Cstring%2CRegex%2Cproc%28RegexMatch%29

I can't think of an elegant way of exposing this kind of functionality in the API, but I probably haven't thought hard enough.

**Comment by timotheecour** [comment]
> I can't think of an elegant way of exposing this kind of functionality in the API, but I probably haven't thought hard enough.

in D, the old API used to be greedy (operate on everything), but the new, recommended way is to operate on 1st element, which is the most flexible, see:
* https://dlang.org/phobos/std_regex.html#.matchFirst
* https://dlang.org/phobos/std_regex.html#.replaceFirst
